deepset_cloud_sdk/workflows/pipeline_client/pipeline_service.py

```diff
@@ -0,0 +1,239 @@
+        async with DeepsetCloudAPI.factory(api_config) as api:
+            service = PipelineService(api)
+            await service.publish_async(self, config)
+
+    def publish_to_deepset(self: PipelineProtocol, config: IndexConfig | PipelineConfig) -> None:
+        """Publish index or pipeline to deepset AI platform synchronously.
+
+        An index is a special type of pipeline with the purpose to preprocess files, preparing them for
+        search and store them in a document store.
+
+        :param config: Configuration for publishing, use either `IndexConfig` or `PipelineConfig`.
+            If publishing an index, the config argument is expected to be of type `IndexConfig`,
+            if publishing a pipeline, the config argument is expected to be of type `PipelineConfig`.
+        """
+        # creates a sync wrapper around the async method since the APIs are async
+        try:
+            loop = asyncio.get_event_loop()
+            should_close = False
+        except RuntimeError:
+            loop = asyncio.new_event_loop()
+            asyncio.set_event_loop(loop)
+            should_close = True
+
+        try:
+            return loop.run_until_complete(publish_to_deepset_async(self, config))
+        finally:
+            if should_close:
+                loop.close()
+
+    def add_method_if_not_exists(cls: type, method_name: str, method: Any, class_name: str) -> None:
+        """Add a method to a class if it doesn't exist.
+
+        :param cls: Class to add the method to
+        :param method_name: Name of the method to add
+        :param method: Method to add
+        :param class_name: Name of the class for logging
+        """
+        if not hasattr(cls, method_name):
+            setattr(cls, method_name, method)
+            logger.debug(f"Successfully added {method_name} method to {class_name} class")
+        else:
+            logger.debug(f"{method_name} method already exists on {class_name} class")
+
+    # Add methods to both Pipeline classes
+    add_method_if_not_exists(HaystackPipeline, "publish_async", publish_to_deepset_async, "Pipeline")
+    add_method_if_not_exists(HaystackPipeline, "publish", publish_to_deepset, "Pipeline")
+    add_method_if_not_exists(HaystackAsyncPipeline, "publish_async", publish_to_deepset_async, "AsyncPipeline")
+    add_method_if_not_exists(HaystackAsyncPipeline, "publish", publish_to_deepset, "AsyncPipeline")
+
+
+class PipelineService:
+    """Handles the publishing of Haystack pipelines and indexes to deepset AI platform."""
+
+    def __init__(self, api: DeepsetCloudAPI) -> None:
+        """Initialize the pipeline service.
+
+        :param api: An initialized DeepsetCloudAPI instance
+        """
+        self._api = api
+        self._yaml = YAML()
+        self._yaml.preserve_quotes = True
+        self._yaml.indent(mapping=2, sequence=2)
+
+    @classmethod
+    async def factory(cls, config: CommonConfig) -> "PipelineService":
```

Comment:
```suggestion
    async def factory(cls, config: CommonConfig) -> PipelineService:
```

from __future__ import annotations